src/main/java/dan200/computercraft/core/computer/ComputerThread.java

```diff
@@ -6,227 +6,304 @@
+                Thread thread = this.thread = new Thread( group, runner, "ComputerCraft-Computer-Runner" + s_counter.getAndIncrement() );
+                thread.setDaemon( true );
+                thread.start();
+            }
+
+            // Execute the task
+            runner.submit( task );
+
+            try
+            {
+                // If we timed out rather than exiting:
+                boolean done = runner.await( 7000 );
+                if( !done )
+                {
+                    // Attempt to soft then hard abort
+                    Computer computer = task.getOwner();
+                    if( computer != null )
+                    {
+                        computer.abort( false );
+
+                        done = runner.await( 1500 );
+                        if( !done )
+                        {
+                            computer.abort( true );
+                            done = runner.await( 1500 );
+                        }
+                    }
+
+                    // Interrupt the thread
+                    if( !done )
+                    {
+                        thread.interrupt();
+                        thread = null;
+                        runner = null;
+                    }
+                }
+            }
+            catch( InterruptedException ignored )
+            {
+            }
+
+            // Re-add it back onto the queue or remove it
+            synchronized( s_taskLock )
+            {
+                if( queue.isEmpty() )
+                {
+                    s_computerTasksActiveSet.remove( queue );
+                }
+                else
+                {
+                    s_computerTasksActive.add( queue );
+                }
+            }
         }
-        
-        synchronized ( m_computerTasksPending )
+    }
+
+    /**
+     * Responsible for the actual running of tasks. It waitin for the {@link TaskRunner#input} semaphore to be
+     * triggered, consumes a task and then triggers {@link TaskRunner#finished}.
+     */
+    private static final class TaskRunner implements Runnable
+    {
+        private final Semaphore input = new Semaphore();
+        private final Semaphore finished = new Semaphore();
+        private ITask task;
+
+        @Override
+        public void run()
         {
-            if( queue.offer( _task ) )
+            try
             {
-                if( !m_computerTasksPending.contains( queue ) )
+                while( true )
                 {
-                    m_computerTasksPending.add( queue );
+                    input.await();
+                    try
+                    {
+                        task.execute();
+                    }
+                    catch( Throwable e )
+                    {
+                        System.out.println( "ComputerCraft: Error running thread." );
+                        e.printStackTrace( System.out );
```

Comment:
This part was just a direct copy of the original implementation :). I think it would probably be safe just to catch `Exception`, though we may also want to catch `OrphanedThread`.